pkg/graveler/graveler.go

```diff
@@ -1640,31 +1640,55 @@ func (g *Graveler) Get(ctx context.Context, repository *RepositoryRecord, ref Re
 		return nil, err
 	}
 
+	var updatedValue *Value
 	if reference.StagingToken != "" {
 		// try to get from staging, if not found proceed to committed
-		value, err := g.getFromStagingArea(ctx, reference.Branch, key)
+		updatedValue, err = g.getFromStagingArea(ctx, reference.Branch, key)
 		if err != nil && !errors.Is(err, ErrNotFound) {
 			return nil, err
 		}
-		if err == nil {
-			if value == nil {
-				// tombstone - the entry was deleted on the branch => doesn't exist
-				return nil, ErrNotFound
-			}
-			return value, nil
+		// tombstone - the entry was deleted on the branch => doesn't exist
+		if err == nil && updatedValue == nil {
+			return nil, ErrNotFound
+		}
+	}
+
+	if updatedValue == nil && reference.CompactedBaseMetaRangeID != "" {
+		updatedValue, err = g.CommittedManager.Get(ctx, repository.StorageNamespace, reference.CompactedBaseMetaRangeID, key)
+		// no need to check for ErrNotFound, since if the key is not found in the compacted base, it will not be found in the committed, and we already checked the staging area
+		if err != nil {
+			return nil, err
 		}
 	}
 
 	var options GetOptions
 	for _, opt := range opts {
 		opt(&options)
 	}
+
+	commitID := reference.CommitID
 	if options.StageOnly {
-		return nil, ErrNotFound
-	}
+		if updatedValue == nil {
+			return nil, ErrNotFound
+		}
 
+		commit, err := g.RefManager.GetCommit(ctx, repository, commitID)
+		if err != nil {
+			return nil, err
+		}
+		committedVal, err := g.CommittedManager.Get(ctx, repository.StorageNamespace, commit.MetaRangeID, key)
+		if err != nil && !errors.Is(err, ErrNotFound) {
+			return nil, err
+		}
+		// the key we found is committed, return not found in staging
+		if diffs := deep.Equal(committedVal, updatedValue); diffs == nil {
+			return nil, ErrNotFound
+		}
+	}
+	if updatedValue != nil {
+		return updatedValue, nil
+	}
```

Comment:
I think this is a bug (so we need to add tests for it 👿  ).
If there's a tombstone in `CompactedBase` I should return it too. I shouldn't need to go and search the value in `Commit`.

Reply:
Cool COol COOl COOL!